woocommerce/woocommerce-gateway-stripe
# Enhance logging of webhook events

Fixes #3328

## Changes proposed in this Pull Request:

- Introduce a method in the WC_Stripe_Logger class for adding log entries of type `debug` and another for entries of type `error`
- When a webhook event is received, log its [type](https://docs.stripe.com/api/events/object#event_object-type) and the entire request body
- When a webhook event fails validation, log the reason behind the validation failure

## Testing instructions

1. Ensure that:
  - Logging is enabled on the Settings tab > Advanced settings > Log error messages, at `siteurl/wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe&panel=settings`
  - Your website is online
  - Webhooks are configured
 2. Go to the webhooks configurations in your stripe dashboard, at https://dashboard.stripe.com/test/webhooks
 3. Add another endpoint pointing to the same URL. You'll have two endpoints pointing to your site
 4. As a shopper, add a product to your cart and place the order
 5. Go to the WC Logs page, at `siteurl/wp-admin/admin.php?page=wc-status&tab=logs`
 6. Open the latest `woocommerce-gateway-stripe` log
 7. Confirm that
   - There are `DEBUG` entries for the `payment_intent.succeeded` and `charge.succeeded` events, with a "Webhook received" message, and another `DEBUG` entries for their request body
   - There are `ERROR` entries with the message "Webhook failed validation. Reason: signature_mismatch" and another `ERROR` entries for their request body 

**Test the "debug" log setting is respected**
1. Disable logging on the Settings tab > Advanced settings > Log error messages, at `siteurl/wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe&panel=settings`
1. As a shopper, add a product to the cart and place the order
1. Go to the WC Logs page, at `siteurl/wp-admin/admin.php?page=wc-status&tab=logs`
1. Open the latest `woocommerce-gateway-stripe` log
1. Confirm no entries were added

---


**Post merge**
